trinity/chains/__init__.py

```diff
@@ -144,6 +148,33 @@ def initialize_database(chain_config: ChainConfig, chaindb: AsyncChainDB) -> Non
             )
 
 
+class TracebackKeeper:
+    """
+    Wrap the given instance, delegating all attribute accesses to it but if any method call raises
+    an exception the traceback is stored in the exception's _tb attribute before it is re-raised.
+    """
+
+    def __init__(self, obj: Any) -> None:
+        self.obj = obj
+
+    def __dir__(self) -> List[str]:
+        return dir(self.obj)
+
+    def __getattr__(self, name: str) -> Any:
+        attr = getattr(self.obj, name)
+        if not inspect.ismethod(attr):
+            return attr
+
+        def wrapper(*args: Any, **kwargs: Any) -> None:
```

Comment:
**maybe** lift this out of the function body to the top level of the module and then this function body would be 

```
if not inspect.ismethod(attr):
    return attr
else:
    return record_traceback_on_error(attr)
```